lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1264 @@
+static const std::string COMP_INPUT_PREFIX = "compin_";
+static const std::string CONST_PREFIX = "const_";
+
+std::error_code InstSynthesis::synthesize(SMTLIBSolver *SMTSolver,
+                                          const BlockPCs &BPCs,
+                                          const std::vector<InstMapping> &PCs,
+                                          Inst *LHS, Inst *&RHS, InstContext &IC,
+                                          unsigned Timeout) {
+  std::error_code EC;
+  RHS = 0;
+
+  // The order is important
+  setCompLibrary();
+  initInputVars(LHS, IC);
+  setDefaultWidth(LHS);
+  addZSTComps(LHS);
+  initOutput(LHS, IC);
+  initComponents(IC);
+  initConstComponents(IC);
+  initLocations();
+
+  N = I.size();
+  M = Comps.size() + N;
+
+  int LHSCost = cost(LHS);
+
+  if (DebugSynthesis)
+    printInitInfo();
+
+  setInvalidWirings();
+
+  // Init a new set of path conditions
+  std::vector<InstMapping> WiringPCs;
+  addConstraints(WiringPCs, IC);
+
+  // Create the main wiring query (aka connectivity contraint)
+  Inst *WiringQuery = getConnectivityConstraint(IC);
+
+  // Initial concrete input set S.
+  // With every new input set that proves a synthesised program is invalid,
+  // we'll have to copy WiringQuery and replace its inputs with the new
+  // concrete values from S
+  std::vector<std::map<Inst *, Inst *>> S;
+
+  // Ask the solver for a couple of arbitrary initial inputs
+  auto InputPCs = PCs;
+  for (unsigned I = 0; I < 4; ++I) {
+    std::vector<Inst *> ModelInsts;
+    std::vector<llvm::APInt> ModelVals;
+    InstMapping Mapping(LHS, IC.createVar(LHS->Width, "foo"));
+    // Negate the query to get a SAT model
+    std::string QueryStr = BuildQuery(BPCs, InputPCs, Mapping,
+                                      &ModelInsts, /*Negate=*/true);
+    bool IsSat;
+    EC = SMTSolver->isSatisfiable(QueryStr, IsSat, ModelInsts.size(),
+                                  &ModelVals, Timeout);
+    if (EC)
+      return EC;
+
+    if (!IsSat)
+      break;
+
+    std::map<Inst *, Inst *> ConcreteInputs;
+    for (unsigned J = 0; J < ModelInsts.size(); ++J) {
+      auto Name = ModelInsts[J]->Name;
+      if (Name.find(INPUT_PREFIX) != std::string::npos) {
+        auto Input = ModelInsts[J];
+        ConcreteInputs[Input] = IC.getConst(ModelVals[J]);
+        Inst *Ne = IC.getInst(Inst::Ne, 1, {Input, IC.getConst(ModelVals[J])});
+        InputPCs.emplace_back(Ne, IC.getConst(APInt(1, true)));
+      }
+    }
+    S.push_back(ConcreteInputs);
+  }
+
+  // Set the maximum number of components
+  int MaxCompNum;
+  if (CmdMaxCompNum >= 0)
+    MaxCompNum = CmdMaxCompNum;
+  else
+    MaxCompNum = Comps.size();
+  if (!IgnoreCost && MaxCompNum >= LHSCost)
+    MaxCompNum = LHSCost-1;
+  // MaxCompNum can be negative when e.g. LHSCost is 0
+  if (MaxCompNum > (int)Comps.size())
+    MaxCompNum = Comps.size();
+  // Iterative synthesis loop with increasing number of components
+  for (int J = 0; J <= MaxCompNum; ++J) {
```

Comment:
Yes, ``MaxCompNum`` becomes negative for the following LHS with cost 0:

```
%0:i8 = var
infer %0
```

MaxCompNum = LHSCost-1;
=> MaxCompNum = -1;

In this case, we don't want to enter the synthesis loop at all.